### PR TITLE
Spatial connections/neighborhood-defining/Label class

### PR DESCRIPTION
**Pull request by dzhu8**
Post-running isort and black on everything

### CODE REVIEW

File: spateo/plotting/static/pairwise.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import matplotlib.patheffects as PathEffects
+import matplotlib.pyplot as plt
+import numpy as np
+import scipy
+from anndata import AnnData
+from matplotlib.collections import PolyCollection
+from matplotlib.ticker import StrMethodFormatter
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+from ...utils import Label, interlabel_connections
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:19]: Looks like the sorting of the imports is not correct? did the isort work properly?

File: spateo/plotting/static/pairwise.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+plt.rcParams["figure.dpi"] = 300
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:21]: spateo has a default figure dpi in the st.configuration.py file, probably no need to include it here

Comment (Xiaojieqiu) [reply spateo/plotting/static/pairwise.py:21]: https://github.com/aristoteleo/spateo-release/blob/3606ba7dda342854781d6e51b9d6a165a67d4a7f/spateo/configuration.py#L505

File: spateo/plotting/static/pairwise.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    figsize: tuple = (8, 8),
+    zero_self_connections: bool = True,
+    normalize_by_self_connections: bool = False,
+    shapes_style: bool = True,
+    max_scale: float = 0.46,
+    colormap: Union[str, dict, "mpl.colormap"] = "Spectral",
+    title_str="connection strengths between types",
+    title_fontsize: float = 12,
+    label_fontsize: float = 12,
+    verbose: bool = True,
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:37]: would be nice to add the following arguments: 

```
    save_show_or_return: Literal["save", "show", "return", "both", "all"] = "show",
    save_kwargs: Optional[dict] = None,
```
to be consistent with other functions

File: spateo/plotting/static/pairwise.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+) -> None:
+    """
+    Plot connections between labels- visualization of how closely labels are colocalized
+
+    Parameters
+    ----------
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:43]: we use google style documentation, so here we will use `args`

File: spateo/plotting/static/pairwise.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    shapes_style : bool, default True
+        If True plots squares, if False plots heatmap
+    max_scale : float, default 0.46
+        Only used for the case that 'shape_style' is True, gives maximum size of square
+    colormap : str, dict, or matplotlib.colormap
+        Specifies colors to use for plotting. If dictionary, keys should be numerical labels corresponding to those
+        of the Label object.
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:50]: the docs here doesn't match with all the arguments. need to update

File: spateo/plotting/static/pairwise.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+from matplotlib.collections import PolyCollection
+from matplotlib.ticker import StrMethodFormatter
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+from ...utils import Label, interlabel_connections
+
+plt.rcParams["figure.dpi"] = 300
+
+
+def plot_connections(
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:24]: I realized that we need to add a legend for the meaning of the box size for this function

File: spateo/plotting/static/pairwise.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+    connections_max = np.amax(connections)
+
+    # Set label colors:
+    if isinstance(colormap, str):
+        cmap = mpl.cm.get_cmap(colormap)
+    else:
+        cmap = colormap
+
+    # If colormap is given, map label ID to points along the colormap. If dictionary is given, instead map
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:72]: instead map `label ID` to the color names? 

need to complete the above sentence

File: spateo/plotting/static/pairwise.py (hunk 6 above)
Comment (Xiaojieqiu) [reply spateo/plotting/static/pairwise.py:24]: Screen Shot 2022-08-28 at 1 22 20 PM
The x-axis ticks need to be on the top

Comment (Xiaojieqiu) [reply spateo/plotting/static/pairwise.py:24]: the ticks on the left axis is good

File: spateo/plotting/static/pairwise.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+            path_effects=[PathEffects.Stroke(linewidth=0.5, foreground="black", alpha=0.8)],
+        )
+
+    ax.set_yticks(np.arange(label.num_labels))
+    if label.str_map is not None:
+        ax.set_yticklabels(
+            label.str_ids,
+            fontsize=label_fontsize,
+            fontweight="bold",
+            path_effects=[PathEffects.Stroke(linewidth=0.5, foreground="black", alpha=0.8)],
```
Comment (Xiaojieqiu) [inline spateo/plotting/static/pairwise.py:170]: make the patheEffects stroke an argument also? when it is None, it uses the above default values but when it is set, it needs to be a list of `PathEffects.Stroke`

File: spateo/utils.py [hunk 9]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+
+            _, _, added_labels = _insert_label(sorted_matched, labels_2.num_labels)
+            relabeled_ids[unmatched_indices] = np.random.choice(added_labels, size=num_extra_labels, replace=False)
+
+        else:
+            logger.error(f"Extra labels assignment method not recognised, should be random or greedy.")
+            raise ValueError(f"Extra labels assignment method not recognised, should be random or greedy.\n")
+
+        if verbose:
+            print(f"\nRelabeled labels: {relabeled_ids}\n")
```
Comment (Xiaojieqiu) [inline spateo/utils.py:397]: may use logger here also?

File: spateo/utils.py [hunk 10]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+        available_labels.remove(label_1)
+
+    # Assign remaining labels (if 2nd has more labels than 1st)
+    if num_extra_labels > 0:
+        unmatched_indices = np.nonzero(relabeled_ids == -1)[0]
+
+        assert num_extra_labels == len(unmatched_indices), (
+            f"number of unmatched label IDs {len(unmatched_indices)} does not match mumber of "
+            f"extra labels in second set {num_extra_labels}.\n"
+        )
```
Comment (Xiaojieqiu) [inline spateo/utils.py:353]: may use logger here also?

File: spateo/utils.py [hunk 11]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+    relabeled_ids = -1 * np.ones((labels_2.num_labels,), dtype=np.int32)
+
+    # Reassign labels
+    for index_1, index_2 in zip(labels_match_1, labels_match_2):
+        label_1 = labels_1.ids[index_1]
+        label_2 = labels_2.ids[index_2]
+        if verbose:
+            print(
+                f"Assigning first set's {label_1} to " f"second set's {label_2}.\n" f"labels_left: {available_labels}"
+            )
```
Comment (Xiaojieqiu) [inline spateo/utils.py:342]: may use logger here also?

File: spateo/utils.py [hunk 12]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+    logger = lm.get_main_logger()
+    max_id = max(labels_1.max_id, labels_2.max_id)
+    num_extra_labels = labels_2.num_labels - labels_1.num_labels
+
+    logger.info(f"Matching {labels_2.num_labels} labels against {labels_1.num_labels} labels.")
+    if verbose:
+        print(
+            f"Matching {labels_2.num_labels} labels against {labels_1.num_labels} labels.\n"
+            f"highest label ID in both is {max_id}.\n"
+        )
```
Comment (Xiaojieqiu) [inline spateo/utils.py:320]: may use logger here also?

File: spateo/utils.py [hunk 13]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+    expanded_ids = ids_zeroed + inserted
+    expanded_ids[1:] += np.cumsum(extra)  # only add to 2nd label and above
+
+    if verbose:
+        print(
+            f"Label ids zerod: {ids_zeroed}.\n"
+            f"{multiple} to be inserted between each id: {inserted}\n"
+            f"{remainder} extra rows to be randomly inserted: {extra}\n"
+            f"New ids: {expanded_ids}"
+        )
```
Comment (Xiaojieqiu) [inline spateo/utils.py:294]: may use logger here also?

File: spateo/utils.py [hunk 14]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+            previous_ptr = indptr[n]
+            current_ptr = previous_ptr + label_count
+            indptr[n + 1] = current_ptr
+
+            if verbose:
+                print(
+                    f"indices for label {label}: {label_indices}\n"
+                    f"previous pointer: {previous_ptr}, "
+                    f"current pointer: {current_ptr}\n"
+                )
```
Comment (Xiaojieqiu) [inline spateo/utils.py:239]: may use logger here also?

File: spateo/utils.py [hunk 15]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+        data = np.ones_like(indices, dtype=np.int32)
+
+        if verbose:
+            print(
+                f"\n--- {self.num_labels} labels, "
+                f"{self.num_samples} samples ---\n"
+                f"initalized {indptr.shape} index ptr: {indptr}\n"
+                f"initalized {indices.shape} indices: {indices}\n"
+                f"initalized {data.shape} data: {data}\n"
+            )
```
Comment (Xiaojieqiu) [inline spateo/utils.py:223]: may use logger here also?

File: spateo/utils.py [hunk 16]
```diff
@@ -30,3 +35,503 @@ def copy_adata(adata: AnnData) -> AnnData:
...
+    str_map : optional dictionary
+        Optional mapping of numerical labels (keys) to strings (values)
+
+    """
+
+    def __init__(
+        self,
+        labels_dense: Union[np.ndarray, list],
+        str_map: Union[None, dict] = None,
+        verbose: bool = True,
```
Comment (Xiaojieqiu) [inline spateo/utils.py:105]: we have a systems level verbose in logger.

File: spateo/plotting/static/pairwise.py (hunk 3 above)
Comment (Xiaojieqiu) [reply spateo/plotting/static/pairwise.py:37]: also we have a system level setup for the verbose in the logger. may need to integrate with that 
```Set the logging level. Various logging level can be setted according to your needs:

DEBUG: useful for dynamo development, show all logging information, including those debugging information

INFO: useful for most dynamo users, show detailed dynamo running information

WARNING: show only warning information

ERROR: show only exception or error information

CRITICAL: show only critical information

%matplotlib inline
from dynamo.dynamo_logger import main_info, LoggerManager
LoggerManager.main_logger.setLevel(LoggerManager.INFO)```

Comment (Xiaojieqiu) [reply spateo/plotting/static/pairwise.py:37]: (here the dynamo logger is used as an example)

File: spateo/plotting/static/pairwise.py (hunk 1 above)
Comment (dzhu8) [reply spateo/plotting/static/pairwise.py:19]: From a glance, the sorting seems consistent with other files in the repository- I think any **import** statements are ordered before any **from** statements, but within each of the import blocks and for blocks all statements are placed in alphabetical order.

**Pull request merged**